include/cantera/numerics/PreconditionerBase.h

```diff
@@ -0,0 +1,135 @@
+
+namespace Cantera
+{
+
+/**
+ * Specifies the preconditioner type used in the integration if any. Not all methods are
+ * supported by all integrators.
+ */
+enum class PreconditionerType {
+    NO_PRECONDITION, //! No preconditioning
+    LEFT_PRECONDITION, //! Left side preconditioning
+    RIGHT_PRECONDITION, //! Right side preconditioning
+    BOTH_PRECONDITION //! Left and right side preconditioning
+};
+
+//! PreconditionerBase serves as an abstract type to extend different preconditioners
+class PreconditionerBase
+{
+public:
+    PreconditionerBase() {}
+
+    virtual const double& operator() (size_t row, size_t col) {
+        throw NotImplementedError("PreconditionerBase::operator()");
+    }
+
+    virtual void operator() (size_t row, size_t col, double value) {
+        throw NotImplementedError("PreconditionerBase::operator()");
+    }
+
+    //! Get preconditioner type for CVODES
+    virtual PreconditionerType preconditionerType() { return PreconditionerType::NO_PRECONDITION; };
+
+    //! Solve a linear system Ax=b where A is the preconditioner
+    //! @param[in] stateSize length of the rhs and output vectors
+    //! @param[in] rhs_vector right hand side vector used in linear system
+    //! @param[out] output output vector for solution
+    virtual void solve(const size_t stateSize, double *rhs_vector, double* output) {
+        throw NotImplementedError("PreconditionerBase::solve");
+    };
+
+    //! Perform preconditioner specific post-reactor setup operations such as factorize.
+    virtual void setup() {
+        throw NotImplementedError("PreconditionerBase::setup");
+    };
+
+    //! Function to reset preconditioner parameters as needed
+    virtual void reset() {
+        throw NotImplementedError("PreconditionerBase::reset");
+    };
+
+    //! This function is called during setup for any processes that need to be completed
+    //! prior to setup functions
+    //! @param network A pointer to the reactor net object associated with the
+    //! integration
+    virtual void initialize(size_t networkSize) {
+        throw NotImplementedError("PreconditionerBase::initialize");
+    };
+
+    //! Print preconditioner contents
+    virtual void printPreconditioner() {
+        throw NotImplementedError("PreconditionerBase::printPreconditioner");
+    };
+
+
+    //! Set gamma used in preconditioning
+    //! @param gamma used in M = I - gamma*J
+    virtual void setGamma(double gamma) {
+        m_gamma = gamma;
+    };
+
+    //! Get gamma used in preconditioning
+    virtual double gamma() {
+        return m_gamma;
+    };
+
+    //! Set the absolute tolerance in the solver outside of the network initialization
+    //! @param atol the specified tolerance
+    virtual void setAbsoluteTolerance(double atol) {
+        m_atol = atol;
+    }
+
+    //! Set dimensions of the preconditioner
+    //! @param dims A pointer to a vector of the dimensions
+    virtual void setDimensions(std::vector<size_t> *dims) {
```

Comment:
the `setDimensions` functions and `getDimensions` functions are actually very early parts of the implementation and at this point I don't believe they are used. It would probably be best to remove them.